driver-core/src/main/java/com/datastax/driver/core/PreparedId.java

```diff
@@ -15,25 +15,125 @@
  */
 package com.datastax.driver.core;
 
+import java.util.List;
+
+import static com.datastax.driver.core.ProtocolVersion.V4;
+
 /**
  * Identifies a PreparedStatement.
  */
+// This class is mostly here to group PreparedStatement data that are need for
+// execution but that we don't want to expose publicly (see JAVA-195)
 public class PreparedId {
-    // This class is mostly here to group PreparedStatement data that are need for
-    // execution but that we don't want to expose publicly (see JAVA-195)
-    final MD5Digest id;
 
-    final ColumnDefinitions metadata;
-    final ColumnDefinitions resultSetMetadata;
+    private volatile MD5Digest id;
+    private volatile ColumnDefinitions variables;
+    private volatile ColumnDefinitions resultSetMetadata;
+    private volatile int[] routingKeyIndexes;
+    private volatile ProtocolVersion protocolVersion;
 
-    final int[] routingKeyIndexes;
-    final ProtocolVersion protocolVersion;
-
-    PreparedId(MD5Digest id, ColumnDefinitions metadata, ColumnDefinitions resultSetMetadata, int[] routingKeyIndexes, ProtocolVersion protocolVersion) {
+    private PreparedId(MD5Digest id, ColumnDefinitions variables, ColumnDefinitions resultSetMetadata, int[] routingKeyIndexes, ProtocolVersion protocolVersion) {
         this.id = id;
-        this.metadata = metadata;
+        this.variables = variables;
         this.resultSetMetadata = resultSetMetadata;
         this.routingKeyIndexes = routingKeyIndexes;
         this.protocolVersion = protocolVersion;
     }
+
+    MD5Digest getId() {
+        return id;
+    }
+
+    /**
+     * Returns the metadata about bound variables in the statement.
+     *
+     * @return the metadata about bound variables in the statement.
+     */
+    ColumnDefinitions getVariables() {
+        return variables;
+    }
+
+    /**
+     * Returns the metadata about columns returned by the statement.
+     *
+     * @return the metadata about columns returned by the statement.
+     */
+    ColumnDefinitions getResultSetMetadata() {
+        return resultSetMetadata;
+    }
+
+    int[] getRoutingKeyIndexes() {
+        return routingKeyIndexes;
+    }
+
+    ProtocolVersion getProtocolVersion() {
+        return protocolVersion;
+    }
+
+    void swap(PreparedId updated) {
```

Comment:
This is a bit hackish but has the advantage of saving us from creating a subinterface for `PreparedStatement`. Note that reads and writes are not atomic (it is possible to read an inconsistent state if state is being updated); we _could_ make them atomic if this is of a concern.